File: src/contracts/app-error.ts
export const APP_ERROR_CODES = [
  "E001",
  "E002",
  "E003",
  "E004",
  "E005",
  "E006",
  "E007",
  "E008",
  "E009",
  "E010",
  "E_INTERNAL",
] as const;

export type AppErrorCode = (typeof APP_ERROR_CODES)[number];

export interface AppError {
  code: AppErrorCode;
  message: string;
  details?: string;
}

const appErrorCodes = new Set<string>(APP_ERROR_CODES);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isAppErrorCode(value: unknown): value is AppErrorCode {
  return typeof value === "string" && appErrorCodes.has(value);
}

export function normalizeIpcError(value: unknown): AppError {
  if (isRecord(value) && isAppErrorCode(value.code) && typeof value.message === "string") {
    const error: AppError = { code: value.code, message: value.message };
    if (typeof value.details === "string") error.details = value.details;
    return error;
  }

  return {
    code: "E_INTERNAL",
    message: typeof value === "string" && value.trim() ? value : "Unexpected application error",
  };
}

File: src-tauri/src/models/app_error.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AppErrorCode {
    E001,
    E002,
    E003,
    E004,
    E005,
    E006,
    E007,
    E008,
    E009,
    E010,
    #[serde(rename = "E_INTERNAL")]
    Internal,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppError {
    pub code: AppErrorCode,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

impl AppError {
    pub fn new(code: AppErrorCode, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            details: None,
        }
    }

    pub fn internal(message: impl Into<String>) -> Self {
        Self::new(AppErrorCode::Internal, message)
    }
}

#[cfg(test)]
mod tests {
    use super::{AppError, AppErrorCode};
    use serde_json::json;

    #[test]
    fn error_serializes_without_empty_details() {
        let value = serde_json::to_value(AppError {
            code: AppErrorCode::E003,
            message: "Invalid link".into(),
            details: None,
        })
        .expect("AppError should serialize");

        assert_eq!(value, json!({ "code": "E003", "message": "Invalid link" }));
    }

    #[test]
    fn internal_error_uses_the_stable_fallback_code() {
        let value = serde_json::to_value(AppError::internal("Unexpected error"))
            .expect("AppError should serialize");

        assert_eq!(
            value,
            json!({ "code": "E_INTERNAL", "message": "Unexpected error" })
        );
    }
}
